### PR TITLE
extract1d updates for IFU data

### PR DESCRIPTION
**Pull request by jemorrison**
This PR is just to show @drlaw1558  what I have so far in setting up the extracct1d data model for miri ifu data
and how I am reading it in.

Fixes [JP-1615](https://jira.stsci.edu/browse/JP-1615)
Fixes [JP-1633](https://jira.stsci.edu/browse/JP-1633)
Fixes [JP-1730](https://jira.stsci.edu/browse/JP-1730)

### CODE REVIEW

**Comment by jemorrison** [comment]
@drlaw1558  made a change to extract1d data model. It now is reading the wavelength array correctly.

File: jwst/datamodels/schemas/mirmrs_extract1d.schema.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      - name: subpixels
+        datatype: int16
+    extract1d_table:
+      title: wavelength varying extraction parameters
+      fits_hdu: X1D
+      datatype:
+      - name: nelem_wl
+        datatype: int16
+      - name: wavelength
+        shape: [3000]
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/datamodels/schemas/mirmrs_extract1d.schema.yaml]: @jdavies-st Thanks with that change I was then able to figure out why the wavelength array was not being read correctly.

[the remark this answers is not in the document]
Comment (jdavies-st) [reply jwst/datamodels/schemas/mirmrs_extract1d.schema.yaml]: 🎉

**Comment by jemorrison** [comment]
@drlaw1558 
Could you look at how I am using the extract1d radius. 
I modified extract.py to read in the extract 1d reference file as FILE_TYPE_TABLE.
In ifu.extrac_ifu I pull in the extract 1d reference data.
I interpolated the reference data to match the ifu wavelength plane.
I had to convert the radius from arc seconds to pixels. I used the routine compute_scale - which 
I think I did correct.
I pass  extract 1d radius to use in ap corr. 
I have just written the first part of the how to use the new apcorr reference files. I got stuck on the radius values of apcor. When I read them in they are very large: here is an example at a particular wavelength (49408.32 24704.16 98816.64]
They are supposed to be in arc seconds so I am thinking they are being read in wrong - do you agree or do I not understand
what the radius array in the apcorr ref file is. 
I am gone this weekend - so I will continue on Monday figuring out how the radius is being read in wrong.

**Comment by drlaw1558** [comment]
@jemorrison I'll take a look at the code and get back to you.  Regarding the weird radius values, my guess is that this may be related to point 2 that I raised on JP-1615.  I.e., assign_wcs.util.compute_scale is returning scale in degrees where convert_size_units assumes input in arcsec, which I found was inflating radii from 3 pixels to 13,000 pixels.

**Comment by drlaw1558** [comment]
Around line 91 of apply_apcorr it looks like I had to hack in 
self.reference[self.size_key] /= 3600.*compute_scale(
to my test copy

**Comment by jemorrison** [comment]
I might of found the error. I will push over new code later today for you to look at concerning the apcor reference file

**Comment by jemorrison** [comment]
@drlaw1558 
Ok I forgot that apcor has a general init that called compute_scale. So that was the problem. I believe I fixed compute_scale 
So now I will move to figure out how to intepolate the apcor function given wavelength and radius information

**Comment by drlaw1558** [comment]
On that latter point, I assumed that we could do something like this:
- The apcor reference info is a 3xN array, with N wavelengths at each of 3 choices of radius.  There is a similar 3xN radius array.
- Extract1d has a 1d flux array with M wavelengths, and a corresponding radius array (i.e., the radius that was used for the extraction) also with length M
- Since we're now trying to do all of the interpolation in one go, we first interpolate the 3xN apcor reference table to 3xM, so that the entries now correspond to the actual wavelengths in use.  I'm sure python has some way to do this automatically, but just doing the interpolation for apcor[0,:], apcor[1,:], and apcor[2,:] individually should work too.  Do the same for the 3xN radius array to make it 3xM.
- Now at each wavelength we interpolate the 3xM apcorr array to a simple M dimensional array, using the input radius at that wavelength from Extract1D in combination with the 3xM array radius to figure out where the interpolation point is in the [:,thiswavelength] dimension. Output is a single vector of length M that corresponds exactly to the spectrum, and can just be multiplied with it.

Does it sounds like that should work?

**Comment by jemorrison** [comment]
@drlaw1558  Take a look at this PR. My brain got a bit twisted into a pretzel but I hope I finally got this correct. No promises. 
This PR included the new datamodels for extract1d, reading the new extract1d and apcorr ref files and using these reference files in extract1d. I added functions in apply_apcor for the Class ApCorrRadial 
Oh also I hope the scaling problem in convert_size_units is fixed by converting the returned degrees in the arc seconds.

**Comment by drlaw1558** [comment]
@jemorrison Looking at the code now, and trying to run pieces of it.  I'm running into a problem with cube_build on this PR branch though (and on master) with 'ValueError: I/O operation on closed file' errors.  Do you know what that's about offhand?

**Comment by drlaw1558** [comment]
A few comments in no particular order:

- extract_ifu now returns radius_match, so it will need to return something similar when running on an EXTENDED source so that it doesn't crash with radius_match undefined

- Looks like the call to compute_scale in extract_ifu is having some problems.  In particular, you're passing in the location in pixels, but it needs to be passed in as a triplet of (RA,DEC,wavelength) of the source location.

- radius_match, inner_bkg_match, outer_bkg_match all look good in extract_ifu

- Looks like the aperture correction reference file that I created put the radii in the wrong order, it should have been small,medium,large apertures in that order but it went medium,small,large.  I've fixed this in a new version.  I've also temporarily made the various aperture correction factors in the reference file be different for the different radii so that we could test that the interpolation was happening properly.

- The call in find_apcorr_func to interp2d looks like it uses the fill_value incorrectly.  If I try to use values outside the radius range bounded by the reference table I get an error that 'extrapolate' can't be converted from string to float.  Looks like interp1d allows this parameter but interp2d doesn't.  I don't think we really need to use 2d interpolation though; since this function is already called at a specific iwave, we should be able to say something like:
temparray=self.apcorr[:,iwave]
fap=interp1d(radius_apcor,temparray,fill_value="extrapolate")
correction=fap(radius_ifu)
This makes it both simpler and allows the extrapolation beyond the bounds of the array to work.  With this change the aperture corrections look good.

Will try to do some more tests later in the week, but it looks like this is getting close.

**Comment by jemorrison** [comment]
@drlaw1558 
Thanks for detailed comments. The test data I am using is extended source data that I hacked and changed the keyword EXTENDED = POINT. So I do not have a target ra, dec. To test this correctly could you put some POINT Source MRS data I can use.

**Comment by drlaw1558** [comment]
Sure; I put the test cube I was using in the Box folder at https://stsci.app.box.com/folder/123396913795 (along with the new apcorr file with different numbers that allows us to test the interpolation)

**Comment by jemorrison** [comment]
@drlaw1558 
On setting up th

**Comment by jemorrison** [comment]
@drlaw1558 @jmuzerolle 
This PR contains the updates to the extract1d to support the IFU extract1d and apcor reference files.
A new  extract 1d data model for MIR MRS was created.  I am going to put in JP ticket about the issue of if the IFU extract1d ref file should be convert to asdf.  But we still need a new extract1d and apcor ref file  for NIRSpec.
This PR cannot be merged as it is. It will break for NIRSpec data. I can add if exp_type  is MIR MRS do blah and
if NIRSPEC do what it used to do (which was wrong - so not really inclined to do that but I could just to get this PR merged). 

David there is a comment on line 408 in the ifu.py code  - I have a question for you there.
I have tried to put #TODO statements in the code were we need to remove/update code once the format of the final extract1d is decided and when NIRSpec has a delivered new ref file.

**Comment by drlaw1558** [comment]
Will take a look.  Just quickly though, now that we've got the structure working it would be good to consider any changes to the reference file model that we want to support long term development.  In particular, there was discussion about moving some of the basic 'how to run this step' info from an extension of the reference file to a calpar file (following the introduction of such files recently).  Will look up details so that we can iterate prior to finalizing the structures.

**Comment by jemorrison** [comment]
@drlaw1558 What is a calpar file ? Do you mean move some of how to run the step into the extract1.cfg ?

**Comment by jemorrison** [comment]
@jmuzerolle 
Do you think you could supply a new NIRSPec extract1d and apcor reference file to support this PR ?
This PR is set up to read an extract1d ref file for IFU that contains an extraction size that varies with wavelength. Also
the apcorr varies with wavelength (and radius). 
Currently the code will fail for NIRSpec data because it is expecting a different reference file. I can put temporary code to
allow it work  using the existing reference files until you can make these new ones. Just let me know what sort of time line you can make the new reference files

**Comment by jmuzerolle** [comment]
@jemorrison I'm hoping to create a draft file later this week.  Do you need it through CRDS, or can I just send it to you directly?  If the former, it may take more than a week to go through the submission process.

**Comment by jemorrison** [comment]
@jmuzerolle for now just send it directly to me. I can get the code updated to read in the new files.

**Comment by jmuzerolle** [comment]
@jemorrison Will do.

**Comment by jemorrison** [comment]
@jmuzerolle @drlaw1558 
Just so you know I am going to take a stab at writing a routine to convert the new extract1d fits files to asdf.
Asdf seems to be the preferred format. I have never written an asdf file so once I do this I will have you look at the resulting
asdf file (and of course I will do this for NIRSPEC once James sends me an tentative new NIRSpec ).

By the way @drlaw1558 I got some clarification from Jonathan on the 'cal pars" concept. There is currently only one of these in CRDS for the spec2 pipeline (pars-spec2pipleine) and it is only for running nirspec data at this time. CRDS folks are in the process of moving files likes this from under the 'instrument' category  in CRDS and making them a their own new category. So you can think about what you might moved once we have these files - but as I understand it - it will be a while before we have these cal par files in CRDS and being used in a consistent manner.

**Comment by jemorrison** [comment]
I have written a py script to convert the MIRI extract1d fits files to asdf. I now have a preliminary asdf file for review. I made this asdf file by using a new data model. For the time being I just called it mirmrs_extract1d_asdf.schema.yaml. After review of this new asdf file, I will replace the current schema written for the fits file mirmrs_extract1d_asdf.schema.yaml with this new one. 
I have put this asdf file on: /grp/jwst/ssb/morrison/EXTRACT1d. The original fits and the converted asdf file in there. I also put the py script I used. The  asdf file is simple and datamodel is simple so it is probably better in the long run than using a fits file.  It took me a long time to come around to asdf file but I am (after much tears) starting to see the benefit.  I am guess that David and James might want to create their reference files using a version of this script themselves rather than creating a fits version and converting it to asdf. But that is a detail we can work out later.
I might of left something out of the asdf file - so let me know. I was not sure how to handle the units - I put these in the meta data of schema (not standard) but I figured the extract1d code needs to know units especially if the units are different between MIRI and NIRSPEC IFU extract 1d reference file.

**Comment by jemorrison** [comment]
@jmuzerolle 
I got your NIRSpec extract1d file. Take a look at the MIRI one in /grp/jwst/ssb/morrison/EXTRACT1d. The first extension contains basic data such as ''subtract_background', "method", "subpxiels".   To make these somewhat complicated FITs be  simpler and line up better with our pipeline I am converting them to asdf. I already did this for MIRI.  Let me give it a shot for NIRSpec. For MIRI is stored all the data as text rather than binary but that might be a rather large file in the case of NIRSpec. Also take a look at the MIRI one - it has AXIS_RATIO and AXIS_PA columns.  Not sure if we need those. 

@drlaw1558 the AXIS_RATIO and AXIS_PA are not being used in the code are they there in case we want to use them in the future ?

**Comment by jmuzerolle** [comment]
@jemorrison Apologies, I forgot to add the information in the first extension.  I can revise the file if you need.
I should note that this reference file is only applicable to point sources; anything extended will have to be manually reprocessed to account for the specific geometries (automated processing should default to extracting the entire cube, as I think it does now?).  For that reason, we don't need AXIS_RATIO or AXIS_PA.

**Comment by drlaw1558** [comment]
The MIRI AXIS_RATIO and AXIS_PA are indeed placeholders for potential future functionality; they are not populated at the moment.  The motivation is that the MRS PSF appears to be elliptical at short wavelengths, and based on what we find during commissioning it may make sense for the aperture to be so as well.

**Comment by jemorrison** [comment]
This is ready for review @jdavies-st @drlaw1558 @jmuzerolle 
Look at JP-1749 for some details on new asdf reference files for IFU data and how to create them

**Comment by jemorrison** [comment]
@jdavies-st @drlaw1558 @jmuzerolle 
This PR is ready for review. The MIRI and NIRSpec teams need to supply new asdf reference files for extract1d and apcorr. Details on code for making these files along with reference files created from the fits files are found in JP-1749.

**Comment by drlaw1558** [comment]
Working on reviewing this; just filed https://github.com/jemorrison/jwst/pull/4/files to fix a couple of typos.  It looks like there may still be issues in the computation of background values for POINT sources observations, details hopefully tomorrow.

**Review by drlaw1558 — APPROVED** [review]
Ok- have now tested this fairly completely, and it looks good to me!

I've tested:
- That both POINT and EXTENDED branches give FLUX and BACKGROUND values as expected for hand-created test data.
- That the applied aperture correction factor is correct, and that the interpolation is done properly if we deliver a table with different values at different radii in future.
- That the above all works out both when running in spec2 and spec3.

I've also prepared the final asdf reference files that I used to test the code.  The delivery of these files to crds will need to be coordinated with the pipeline update.

**Comment by jemorrison** [comment]
@jdavies-st 
There are some travis tests failing in this PR. 
Latest dependency versions w/coverage and SDP dependencies in requirements seem to both indicate test/test_wcs.py is failing but when I run pytest on it in my local branch it is fine. Is this failing because I need to upgrade to python 3.9  in my environment ?

**Comment by jdavies-st** [comment]
That failure is unrelated (due to a new release of `gwcs` and our dev dependency tests getting screwed up a week ago), and was just fixed in #5483 (and #5482).

**Comment by jdavies-st** [comment]
I've restarted the failing Travis jobs.  👍

**Comment by jemorrison** [comment]
@jdavies-st  when you get a chance can you review this change - in particular the new data models

**Comment by drlaw1558** [comment]
Did we get the failing tests issue sorted out so that we can look into merging this PR?

**Comment by jemorrison** [comment]
@jdavies-st  could you review this code. It uses datamodel updates so someone with the knowledge should review it. 
It is about ready to be merged. We are getting the new reference files ready to be put in the CRDS then all that remains is
for this to be reviewed by a SCSB member. If you want to suggest someone else to do that - ok

**Comment by jemorrison** [comment]
closing - opened a clean new version #5506

**Pull request closed without merging**